_P164_MSswitch.ino

```diff
@@ -0,0 +1,169 @@
+      Device[deviceCount].Type = DEVICE_TYPE_SINGLE;
+      Device[deviceCount].Ports = 0;
+      Device[deviceCount].Custom = true;
+      Device[deviceCount].TimerOption = false;
+      Device[deviceCount].SendDataOption = false;
+      break;
+    }
+
+    case PLUGIN_GET_DEVICENAME:
+    {
+      string = F(PLUGIN_NAME_164);
+      break;
+    }
+
+    case PLUGIN_GET_DEVICEVALUENAMES:
+    {
+      strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[0], PSTR(PLUGIN_VALUENAME1_164));
+      break;
+    }
+
+    case PLUGIN_WEBFORM_LOAD:
+    {
+      addRowLabel(F("433MHZ Data Pin"));
+      addPinSelect(false, F("taskdevicepin1"), CONFIG_PIN1);
+      addFormNumericBox(F("Switch Address"), F("p164_address"), PCONFIG_LONG(0),0,67108863);
+
+      if (CONFIG_PIN1>-1) {
+        addFormSubHeader(F("Try it"));
+        String options[17] = { F("CH 1"),F("CH 2"),F("CH 3"),F("CH 4"),F("CH 5"),F("CH 6"),F("CH 7"),F("CH 8"),F("CH 9"),F("CH 10"),F("CH 11"),F("CH 12"),F("CH 13"),F("CH 14"),F("CH 15"),F("CH 16"),F("GROUP") };
+        int optionValues[17] = { 0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16 };
+        String sOptions[2] = { F("On"),F("Off") };
+        int sOptionsValues[2] = {1,0};
+
+        addFormSelector(F("Channel"), F("channel"), 17, options, optionValues, 0);
+        addFormSelector(F("State"), F("state"), 2, sOptions, sOptionsValues, 0);
+        addFormCheckBox(F("Send on submit"), F("send"), false);
+      }
+      success = true;
+      break;
+    }
+
+    case PLUGIN_WEBFORM_SAVE:
+    {
+      PCONFIG_LONG(0) = getFormItemInt(F("p164_address"));
+      SocketAddress = PCONFIG_LONG(0);
+      if (isFormItemChecked(F("send"))) {
+        noInterrupts();
+        log164 = F("Sending: ");
+        noInterrupts();
+        if (getFormItemInt(F("state"))) {
+          if (getFormItemInt(F("channel"))==16) {
+            Plugin_164_msswitch->groupOn(SocketAddress);
+            log164 += F("GroupOn");
+          } else {
+            Plugin_164_msswitch->channelOn(SocketAddress,getFormItemInt(F("channel")));
+            log164 += F("channelOn: ");
+            log164 += getFormItemInt(F("channel"));
+            log164 += F(" Address: ");
+            log164 += SocketAddress;
+            log164 += F(" PIN: ");
+            log164 += CONFIG_PIN1;
+          }
+        } else {
+          if (getFormItemInt(F("channel"))==16) {
+            Plugin_164_msswitch->groupOff(SocketAddress);
+            log164 += F("GroupOff");
+          } else {
+            Plugin_164_msswitch->channelOff(SocketAddress,getFormItemInt(F("channel")));
+            log164 += F("channelOff: ");
+            log164 += getFormItemInt(F("channel"));
+          }
+        }
+        interrupts();
+        addLog(LOG_LEVEL_DEBUG, log164);
+      }
+      success = true;
+      break;
+    }
+
+    case PLUGIN_INIT:
+    {
+      if (!Plugin_164_msswitch)
+      {
+        Plugin_164_msswitch = new Socketswitch(CONFIG_PIN1);
```

Comment:
You have implemented a `PLUGIN_INIT` but no `PLUGIN_EXIT` in which the pointer must be deleted.
This means this allocation may lead to a memory leak when the plugin is removed.
Also a change of pin settings now need a reboot of the ESP to activate.